### ISSUE TITLE
ImageView flicker when app Re-enter foreground

### ISSUE THREAD

**Issue by weiminghuaa** [rs/SDWebImage#1695]
### New Issue Checklist
- [x] I have read and understood the [CONTRIBUTING guide](https://github.com/rs/SDWebImage/blob/master/.github/CONTRIBUTING.md)
- [x] I have read the [Documentation](http://cocoadocs.org/docsets/SDWebImage/)
- [x] I have searched for a similar issue in the [project](https://github.com/rs/SDWebImage/issues) and found none
### Issue Info

| Info | Value |
| --- | --- |
| Platform Name | ios |
| Platform Version | 10.0 |
| SDWebImage Version | 3.8.2 |
| Integration Method | cocoapods |
| Xcode Version | Xcode 8 |
| Repro rate | all the time (100%) |
### Issue Description and Steps

SDWebImage will clear memory cache when app enter background. So, when app enter foreground, scroll tableView to refresh, and then reloadData, imageView in cell will flicker, even though data haven't changed.

`[cell.toolsImageView sd_setImageWithURL:[NSURL URLWithString:toolsModel.iconUrl] placeholderImage:[UIImage imageNamed:@"morentu_circle"] options:SDWebImageRetryFailed];`

Flicker steps around reloadData:
1. show image before reloadData;
2. reloadData and imageView reset image with URL (Same URL);
3. change to show placeholderImage;
4. find image in SD memory cache, and not found;
5. find image in SD disk cache, get it async;
6. load disk image complete and show.

**Comment by weiminghuaa** [comment]
Is like This?
`- (void)sd_setImageWithURL:(NSURL *)url placeholderImage:(UIImage *)placeholder options:(SDWebImageOptions)options progress:(SDWebImageDownloaderProgressBlock)progressBlock completed:(SDWebImageCompletionBlock)completedBlock {
    
    if (!(options & SDWebImageRefreshCached) && [[self sd_imageURL].absoluteString isEqualToString:url.absoluteString]) {
        return;
    }
    .....
}`

**Comment by weiminghuaa** [comment]
`- (void)sd_setImageWithPreviousCachedImageWithURL:(NSURL *)url placeholderImage:(UIImage *)placeholder options:(SDWebImageOptions)options progress:(SDWebImageDownloaderProgressBlock)progressBlock completed:(SDWebImageCompletionBlock)completedBlock {
    NSString *key = [[SDWebImageManager sharedManager] cacheKeyForURL:url];
    UIImage *lastPreviousCachedImage = [[SDImageCache sharedImageCache] imageFromDiskCacheForKey:key];
 
    [self sd_setImageWithURL:url placeholderImage:lastPreviousCachedImage ?: placeholder options:options progress:progressBlock completed:completedBlock];   
}`

**Comment by weiminghuaa** [comment]
Is like This? @bpoplauschi @rs @CavalcanteLeo @KanybekMomukeyev @bmurmistro @mythodeia   @andodelauni @heistings
    
    `- (void)sd_setImageWithURL:(NSURL *)url placeholderImage:(UIImage *)placeholder options:(SDWebImageOptions)options progress:(SDWebImageDownloaderProgressBlock)progressBlock completed:(SDWebImageCompletionBlock)completedBlock {
                [self sd_cancelCurrentImageLoad];
    
               NSURL *oldURL = [self sd_imageURL];
    
              objc_setAssociatedObject(self, &imageURLKey, url, OBJC_ASSOCIATION_RETAIN_NONATOMIC);

             if (!(options & SDWebImageDelayPlaceholder) && (oldURL == nil || ![oldURL.absoluteString isEqualToString:url.absoluteString])) {
                      dispatch_main_async_safe(^{
                             self.image = placeholder;
                      });
              }
         .....
    }`

**Comment by bpoplauschi** [comment]
@weiminghuaa clearing up the memory cache is a feature of SDWebImage. I don't think we will remove it. You can try to find a way around this

**Comment by weiminghuaa** [comment]
@bpoplauschi  Thank you very much for you reply. I knew this feature. The question has troubled me for long time. Can I ask you: how to avoid this?

 Beacuse the memory clearing up, and cell reuse not follow the fixed order，like this: 

`1st loadData

    2016-09-30 15:55:28.764 MYAPP[12950:384498] PosterCell-0: <PosterCell: 0x7fdd3c57cf90>

    2016-09-30 15:55:28.782 MYAPP[12950:384498] PosterCell-1: <PosterCell: 0x7fdd3c4cc0c0>

    2016-09-30 15:55:28.807 MYAPP[12950:384498] PosterCell-2: <PosterCell: 0x7fdd3c5815e0>

2nd reloadData

    2016-09-30 15:55:28.959 MYAPP[12950:384498] PosterCell-0: <PosterCell: 0x7fdd3c5815e0>

    2016-09-30 15:55:28.961 MYAPP[12950:384498] PosterCell-1: <PosterCell: 0x7fdd3c57cf90>

    2016-09-30 15:55:28.962 MYAPP[12950:384498] PosterCell-2: <PosterCell: 0x7fdd3c4cc0c0>`

So, when app Re-enter foreground, the imageView's image is right. The reload, the imageView set placehold first, and load image from disk, at last, set the right image again. That cause flicker when cpu is lower.

**Comment by weiminghuaa** [comment]
`- (void)sd_setImageWithURL:(nullable NSURL *)url
          placeholderImage:(nullable UIImage *)placeholder
                   options:(SDWebImageOptions)options
                  progress:(nullable SDWebImageDownloaderProgressBlock)progressBlock
                 completed:(nullable SDExternalCompletionBlock)completedBlock {
    if (![self.image.accessibilityIdentifier isEqualToString:@"placeholder"]) {
        NSString *key = [[SDWebImageManager sharedManager] cacheKeyForURL:self.sd_imageURL];
        if (![[SDImageCache sharedImageCache] imageFromMemoryCacheForKey:key]) {
            [[SDImageCache sharedImageCache] storeImage:self.image forKey:key toDisk:NO completion:nil];
        }
    }
    
    placeholder.accessibilityIdentifier = @"placeholder";
    
    [self sd_internalSetImageWithURL:url
                    placeholderImage:placeholder
                             options:options
                        operationKey:nil
                       setImageBlock:nil
                            progress:progressBlock
                           completed:completedBlock];
}`

This way may fix the problem. Right?

**Comment by bpoplauschi** [comment]
It may work. Let me know

**Comment by weiminghuaa** [comment]
Sometimes it work, sometimes not. Beacuse chaotic reuse when reload. I decided load images before reload.

```
          dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
            
            for (XRKHomeToolsModel *toolsModel in _tools) {
                NSString *key = [[SDWebImageManager sharedManager] cacheKeyForURL:[NSURL URLWithString:toolsModel.iconUrl]];
                [[SDImageCache sharedImageCache] imageFromCacheForKey:key];
            }
            
            dispatch_async(dispatch_get_main_queue(), ^{
                [self.toolCollectionView reloadData];
            });
        });
```

**Comment by weiminghuaa** [comment]
Sometimes not work, beacuse chaotic reuse cell when reloadData

**Comment by weiminghuaa** [comment]
The images is in memory when reenter foreground. But memory cache in SDImageCache is cleared. Is there any way to find image from memory?

**Comment by bpoplauschi** [comment]
I don't know if we can do more here. Closing. Feel free to reopen or continue the conversation

**Comment by swapnil-vuclip** [comment]
I have the same related issue mentioned by @weiminghuaa , 
I am using SDWebImage+WebP to load my webp images from my server in list of collection in UItableView 

Every time I re-enter foreground imageView in cell will re-fetch images from disk(mean while image view will be blank) and reload UIImageview with same image,  even though data haven't changed. 

But When I used jpg format with the same code base it will work as I expected, their is black out on re-enter foreground it will shows the cached images. but not with webp   
Code I used :

            [videoV.imageViewVideo sd_setImageWithURL:[NSURL URLWithString:imageUrl]
                                     placeholderImage:nil options:SDWebImageRefreshCached|SDWebImageRetryFailed ];

any idea?
